README.md
# blockshead

Simple 2D shooter game written in Python using Pygame.

# Roadmap

See GitHub issues.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## [0.2.0] - 2023-01-08

### Bug Fixes

- Mirror zombie images
- Actually calculate zombie direction
- Camera movement offsets for devil attacks
- Devil cooldown
- Make zombie spawning gradual
- Number of spawned zombies
- Collision system
- Fake walls
- Planting fake walls

### Documentation

- Update pygame roadmap

### Features

- Start implementing map movement
- Implement map movement; enhance uzi animation
- Automatically change to pistol if out of ammo
- Implement messages
- Implement double and quad ammo
- Add new audio
- Start implementing devil
- Implement devils, use rectangular hitboxes
- Only allow for compass directions for the zombies
- Calculate FPS as a moving average over the past 10 frames
- Create healthbox when devil dies
- End screen and new game
- Implement devil attacks
- Spawn zombies from both directions
- Start implementing fake walls
- Implement recoil

### Refactor

- Move character speeds to game_config
- Use logging for stdout messages

## [0.1.0] - 2022-12-18

### Bug Fixes

- Window size
- Character direction on keyup events
- Collision constants
- Overlapping spawning of zombies
- Collision management
- Offsets in coordinate system
- Collision system
- Zombie spawning
- Float-valued ammo
- Adjust parameters

### Documentation

- Mention pygame
- Complete sentence
- Update pygame roadmap
- Update pygame roadmap

### Features

- Implement main character drawing
- Move character
- Implement character image drawing
- Implement zombie image drawing
- Implement zombie movement
- Start implementing weapons
- Play sound when shooting
- Window title
- Multiple zombies
- Draw score
- Implement pause functionality
- Implement complex zombie movement
- Zombie attacking
- Health mechanics and blood marks
- Implement health bar
- Implement fake walls
- Decay blood marks on levelup
- Draw pistol shots
- Implement uzi
- Implement uzi
- Make max health configurable
- Implement shotgun
- Add audio for shotgun

### Miscellaneous Tasks

- Merge branch 'pygame' of gitlab.com:ninpnin/blockshead into pygame
- Remove debug messages

### Refactor

- Start moving towards pygame
- Use float-based coordinate system
- Moving logic
- Game state dict
- Delete dead code

<!-- generated by git-cliff -->


app.py
import logging
import sys
logging.basicConfig(level=logging.DEBUG, stream=sys.stdout)
from tkinter import W
from blockshead.gamestate import *
from blockshead.characters import Blockshead, Zombie, Devil
from blockshead.objects import *
import pygame
from pygame import mixer
import math
from numpy.random import choice as np_choice

def initialize_game():
    logging.info("initialize game")
    pygame.init()
    pygame.display.set_caption('Blockshead')
    game_config = GameConfig(canvas=None, width=1000, height=750)
    screen = pygame.display.set_mode([game_config.width, game_config.height])
    game_state = GameState()
    game_state.blockshead = Blockshead(game_config)
    return game_config, game_state, screen

def draw_pause_screen(window, game_config):
    window.fill(game_config.background_color)
    font = pygame.font.SysFont(None, 36)
    img = font.render("Paused. Press 'P' to continue", True, (100,100,100))
    window.blit(img, (20, 20))
    
def draw_stats(window, game_state, game_config):
    # Left corner title
    font = pygame.font.SysFont(None, 36)
    img = font.render(f'Level: {game_state.level}, Points: {game_state.score}', True, (100,100,100))
    window.blit(img, (20, 20))
    img = font.render(f'Weapon: {game_state.blockshead.weapon}', True, (100,100,100))
    window.blit(img, (20, 60))
    img = font.render(f'Ammo: {game_state.blockshead.ammo()}', True, (100,100,100))
    window.blit(img, (20, 100))

    # Right corner title
    font = pygame.font.SysFont(None, 36)
    img = font.render(f'Multiplier: {game_state.multiplier:.2f}', True, (100,100,100))
    window.blit(img, (game_config.width - 200, 20))

    # Health bar
    health = game_state.blockshead.health
    healthbar_width = int(100 * health/game_config.max_health)
    start_x = - 40 + game_state.blockshead.x - game_state.offset_x
    start_y = game_state.blockshead.y - 70 - game_state.offset_y
    pygame.draw.rect(window, (255,0,0), (start_x, start_y, 100, 10))
    pygame.draw.rect(window, (0,128,0), (start_x, start_y, healthbar_width, 10))

def draw_messages(window, game_state, game_config):
    # Left corner title
    game_state.messages = [(m, t -1) for m,t in game_state.messages if t >= 1]
    x, y = game_config.width // 2 - 200, 20
    for m, _ in game_state.messages:
        font = pygame.font.SysFont(None, 36)
        img = font.render(m, True, (100,100,100))
        window.blit(img, (x, y))
        y += 40

def add_zombies(game_config, game_state, window):
    x = random.randrange(40, game_config.width // 20)
    if random.choice([True, False]):
        x = game_config.width - x
    y = random.randrange(40, game_config.height -40)

    if game_state.level_zombies >= 1:
        zombie = Zombie(window, game_config, init_coords=(x,y))
        if not zombie._check_collisions(game_state, game_config):
            logging.debug("add zombie")
            game_state.zombies.append(zombie)
            game_state.level_zombies -= 1
        else:
            logging.warning("do not add zombie due to collision")

    devil_rate = 0.1
    if np_choice([True, False], p=[devil_rate, 1.0 - devil_rate]):
        if game_state.level_devils >= 1:
            devil = Devil(window, game_config)
            if not devil._check_collisions(game_state, game_config):
                game_state.devils.append(devil)
                game_state.level_devils -= 1
        
def new_level(game_config, game_state, window):
    logging.info("new level")
    game_state.number_of_zombies += 2
    game_state.level_zombies = game_state.number_of_zombies
    game_state.level_devils = 1 + game_state.number_of_zombies // 10
    logging.info(f"{game_state.level_zombies} zombies and {game_state.level_devils} devils at this level")
    game_state.zombies = []
    
    healthbox_x = random.randrange(0, game_config.width)
    healthbox_y = random.randrange(0, game_config.height)
    healthbox = Healthbox(game_state, healthbox_x, healthbox_y)
    game_state.healthboxes.append(healthbox)
    
    for blood_mark in game_state.blood_marks:
        blood_mark.levelup()
        
    game_state.level += 1
    game_state.messages.append((f"Level {game_state.level}", 180))
    
    newlevel_audio = pygame.mixer.Sound('audio/new-level.mp3')
    mixer.Channel(1).play(newlevel_audio)
    #mixer.music.play()

    return game_config, game_state, window

def handle_keys(event, window, game_config, game_state):
    if event.type == pygame.KEYDOWN:
        if not game_state.paused:
            # Game controls
            if event.key == pygame.K_LEFT:
                game_state.blockshead.x_vel = -1
                game_state.blockshead.direction = Direction.LEFT
            if event.key == pygame.K_RIGHT:
                game_state.blockshead.x_vel = 1
                game_state.blockshead.direction = Direction.RIGHT
            if event.key == pygame.K_UP:
                game_state.blockshead.y_vel = -1
                game_state.blockshead.direction = Direction.UP
            if event.key == pygame.K_DOWN:
                game_state.blockshead.y_vel = 1
                game_state.blockshead.direction = Direction.DOWN
                
            if event.key == pygame.K_SPACE:
                shot = game_state.blockshead.fire_gun(window, game_config, game_state)
                if shot is not None:
                    game_state.shots.append(shot)

            weapon_keys = [pygame.K_1, pygame.K_2, pygame.K_3, pygame.K_4, pygame.K_5]
            weapon_keys += [pygame.K_6, pygame.K_7, pygame.K_8, pygame.K_9, pygame.K_0]
            if event.key in weapon_keys:
                index = weapon_keys.index(event.key)
                if index < len(game_state.available_weapons):
                    game_state.blockshead.weapon = game_state.available_weapons[index]
                else:
                    logging.warning(f"Weapon {index} not available")
                #game_state.available_weapons
        # Pause controls
        if event.key == pygame.K_p:
            game_state.paused = not game_state.paused
            
    elif event.type == pygame.KEYUP:
        if event.key in [pygame.K_LEFT, pygame.K_RIGHT]:
            game_state.blockshead.x_vel = 0
            if game_state.blockshead.y_vel > 0:
                game_state.blockshead.direction = Direction.DOWN
            elif game_state.blockshead.y_vel != 0:
                game_state.blockshead.direction = Direction.UP

        elif event.key in [pygame.K_UP, pygame.K_DOWN]:
            game_state.blockshead.y_vel = 0
            if game_state.blockshead.x_vel > 0:
                game_state.blockshead.direction = Direction.RIGHT
            elif game_state.blockshead.x_vel != 0:
                game_state.blockshead.direction = Direction.LEFT
                
def draw_screen(window, characters, game_state, debug=True):
    """
    Draw characters and their debug circle
    """
    for c in characters:
        img = c.get_image()
        c_coords = c.get_coordinates()
        c_coords = c_coords[0] - game_state.offset_x,  c_coords[1] - game_state.offset_y
        x = c_coords[0] - img.get_width() // 2
        y = c_coords[1] - img.get_height() // 2
        window.blit(img, (x,y))

        if debug:
            x_hitbox_start = c_coords[0] - c.width // 2
            y_hitbox_start = c_coords[1] - c.height // 2
            x_hitbox_end = c.width
            y_hitbox_end = c.height
            pygame.draw.rect(window, (0,255,0), (x_hitbox_start, y_hitbox_start, x_hitbox_end, y_hitbox_end), 3)

def draw_shots(window, game_state):
    for s in game_state.shots:
        s.draw(window, game_state)
        
def update_weapons(game_config, game_state):
    new_weapons, new_ammo = [], dict()
    for weapon, threshold in game_config.weapons.items():
        if game_state.multiplier > threshold and weapon not in game_state.available_weapons:
            logging.info(f"new weapon: {weapon}")
            game_state.messages.append((f"New weapon: {weapon}", 180))
            new_weapons.append(weapon)
            new_ammo[weapon] = game_config.ammo[weapon]
    
    return new_weapons, new_ammo

def update_ammo(game_config, game_state, multiplier=2.0, multiplier4=8.0):
    max_ammo = {w: a for w, a in game_config.ammo.items() if w in game_state.available_weapons}
    for weapon, threshold in game_config.weapons.items():
        if weapon in game_state.available_weapons and game_state.multiplier > threshold * multiplier:
            if type(max_ammo.get(weapon)) in [float, int]:
                max_ammo[weapon] = game_config.ammo[weapon] * 2
                if max_ammo[weapon] > game_state.max_ammo.get(weapon):
                    game_state.messages.append((f"{weapon}: Double ammo", 180))
        if weapon in game_state.available_weapons and game_state.multiplier > threshold * multiplier4:
            if type(max_ammo.get(weapon)) in [float, int]:
                max_ammo[weapon] = game_config.ammo[weapon] * 4
                if max_ammo[weapon] > game_state.max_ammo.get(weapon):
                    game_state.messages.append((f"{weapon}: Quad ammo", 180))

    return max_ammo

def draw_end_screen(game_state, window):
    window.fill(game_config.background_color)
    font = pygame.font.SysFont(None, 36)

    text1 = f"Game over!"
    img1 = font.render(text1, True, (100,100,100))
    window.blit(img1, (20, 20))

    text2 = f"Final Score: {math.ceil(game_state.score)}, Final Level: {game_state.level -1}"
    img2 = font.render(text2, True, (100,100,100))
    window.blit(img2, (20, 60))

    text3 = f"Press R to start a new game. Press any other key to quit"
    img3 = font.render(text3, True, (100,100,100))
    window.blit(img3, (20, 100))

def end_loop(game_state, window, clock):
    logging.info("Enter end loop")
    draw_end_screen(game_state, window)
    pygame.display.update()
    while True:
        dt = clock.tick(60)
        for event in pygame.event.get():
            if event.type == pygame.QUIT:
                return False
            elif event.type == pygame.KEYDOWN:
                return event.key == pygame.K_r

def main_loop(game_config, game_state, window, clock, levelup=False):
    game_config, game_state, window = new_level(game_config, game_state, window)
    
    for _ in range(2):
        wall = Fakewall(game_config)
        game_state.fakewalls.append(wall)
    
    fps = []
    #game_state.messages.append(("ebuns", 100))
    while True:
        if len(game_state.zombies) == 0 and game_state.level_zombies == 0 and len(game_state.devils) == 0:
            game_config, game_state, window = new_level(game_config, game_state, window)
        dt = clock.tick(60)
        fps = fps[-10:]
        fps.append(1000/ dt)

        title_str = f"blockshead – {np.mean(fps):.0f} fps"
        pygame.display.set_caption(title_str)

        # Handle events
        for event in pygame.event.get():
            if event.type == pygame.QUIT:
                return
            elif event.type == pygame.KEYDOWN or event.type == pygame.KEYUP:
                handle_keys(event, window, game_config, game_state)

        if not game_state.paused:
            # Draw background
            window.fill(game_config.background_color)
            
            # Move characters
            game_state.blockshead.move(game_config, game_state)
            for zombie in game_state.devils + game_state.zombies:
                zombie.move(window, game_config, game_state)
                zombie.contact(game_state)

            # Calculate interactions, attacks
            for shot in game_state.shots:
                shot.update(game_config, game_state)

            for healthbox in game_state.healthboxes:
                healthbox.update(game_config, game_state)

            game_state.zombies = [z for z in game_state.zombies if z.health >= 1]
            game_state.devils = [z for z in game_state.devils if z.health >= 1]
            game_state.shots = [s for s in game_state.shots if s.lifetime >= 0]
            game_state.healthboxes = [b for b in game_state.healthboxes if b.active]
            game_state.fakewalls = [b for b in game_state.fakewalls if b.health >= 0]
            game_state.blood_marks = [b for b in game_state.blood_marks if b.level_lifetime >= 0]
            game_state.multiplier = max(game_state.multiplier - game_config.multiplier_step, 1.0)
            new_weapons, _ = update_weapons(game_config, game_state)
            game_state.available_weapons += new_weapons
            game_state.max_ammo = update_ammo(game_config, game_state)
            if len(new_weapons) >= 1:
                logging.debug(f"New weapons {new_weapons}")
                for w in new_weapons:
                    game_state.blockshead.ammo_dict[w] = game_state.max_ammo[w]
                logging.debug(f"max ammo: {game_state.max_ammo}")
            
            if np_choice([True, False], p=[game_config.zombie_spawn_rate, 1.0 - game_config.zombie_spawn_rate]):
                add_zombies(game_config, game_state, window)
            
            # Draw characters and objects
            # NOTE: Order matters here
            drawables = game_state.blood_marks + game_state.fakewalls + game_state.healthboxes
            drawables = drawables + [game_state.blockshead] + game_state.zombies + game_state.devils
            draw_screen(window, drawables, game_state)
            draw_stats(window, game_state, game_config)
            draw_shots(window, game_state)
            draw_messages(window, game_state, game_config)
            
            # Calculate map offset
            ## X-axis
            if game_state.blockshead.x - game_state.offset_x >= game_config.width * 0.8:
                game_state.offset_x = game_state.blockshead.x - game_config.width * 0.8
            elif game_state.blockshead.x - game_state.offset_x <= game_config.width * 0.2:
                game_state.offset_x = game_state.blockshead.x - game_config.width * 0.2
            ## Y-axis
            if game_state.blockshead.y - game_state.offset_y >= game_config.height * 0.8:
                game_state.offset_y = game_state.blockshead.y - game_config.height * 0.8
            elif game_state.blockshead.y - game_state.offset_y <= game_config.height * 0.2:
                game_state.offset_y = game_state.blockshead.y - game_config.height * 0.2
        else:
            draw_pause_screen(window, game_config)
        
        if game_state.blockshead.health <= 0:
            return
        
        pygame.display.update()
        

if __name__ == "__main__":
    clock = pygame.time.Clock()
    start_new_game = True
    while start_new_game:
        game_config, game_state, window = initialize_game()
        main_loop(game_config, game_state, window, clock)
        start_new_game = end_loop(game_state, window, clock)

blockshead/gamestate.py
from dataclasses import dataclass

@dataclass
class GameConfig:
    canvas: bool
    background_color = "#EBDCC7"
    font = ('Helvetica','30','bold')
    weapons = {"pistol": 1, "uzi": 3, "shotgun": 5, "fake walls": 1}
    ammo = {"uzi": 100, "shotgun": 20, "fake walls": 10}
    multiplier_step = 0.0002
    blockshead_speed = 2.0
    zombie_speed = 0.5
    devil_speed = 0.75
    max_health = 200
    map_width = 2000
    map_height = 1500
    zombie_spawn_rate = 0.01
    height: int
    width: int

@dataclass
class GameState:
    direction = 1
    number_of_zombies = 5
    total_devil_attacks = 0
    blood_marks = 0
    number_of_mines = 0
    paused = False
    multiplier = 1
    level_zombies = 1
    available_weapons = ["pistol"]
    max_ammo = {"pistol": "infinity"}
    shots = []
    mines = []
    zombies = []
    devils = []
    devil_attacks = []
    blood_marks = []
    healthboxes = []
    fakewalls = []
    messages = []
    score: int = 0
    level: int = 0
    offset_x = 0
    offset_y = 0

@dataclass
class Shots:
    direction: int
    x: int
    y: int

blockshead/characters.py
import pygame
import random
import numpy as np
from .physics import check_collision
from .objects import Direction
from .objects import Pistol, Blood, Uzi, Shotgun, DevilAttack, Healthbox, Fakewall
import math
import logging

class Blockshead(object):
    def __init__(self, game_config):
        self.images = {}
        self.images[Direction.UP] = pygame.image.load("images/blockshead/bhup.png")
        self.images[Direction.DOWN] = pygame.image.load("images/blockshead/bhdown.png")
        self.images[Direction.LEFT] = pygame.image.load("images/blockshead/bhleft.png")
        self.images[Direction.RIGHT] = pygame.image.load("images/blockshead/bhright.png")
        self.width = 70
        self.height = self.width
        
        self.x = random.randrange(self.width, game_config.width - self.width)
        self.y = random.randrange(self.height, game_config.height - self.height)
        self.direction = Direction.UP
        self.x_vel = 0
        self.y_vel = 0
        self.health = game_config.max_health # +5 health is added at the beginning of every level
        self.weapon = "pistol"
        self.ammo_dict = {"pistol": "Infinite"}
        self.pause = False
        self.bonus_score = 0
        self.pistol_range = 150 # the range of the pistol in pixels
        self.mine_count = 0 # how many mines are left
        self.bullet_images = []
        self.cooldown = 0
        self.speed = game_config.blockshead_speed
    
    def ammo(self):
        return self.ammo_dict.get(self.weapon, 0)
    
    def get_image(self):            
        return self.images[self.direction]

    def _check_collisions(self, game_state):
        for zombie in game_state.zombies + game_state.fakewalls:
            if check_collision(self, zombie):
                return True
        return False

    def move(self, game_config, game_state):
        # game area boundaries
        if (self.x >= game_config.width) and self.x_vel > 0:
            self.x_vel = 0
        elif self.x <= 0 and self.x_vel < 0:
            self.x_vel = 0

        if (self.y >= game_config.height) and self.y_vel > 0:
            self.y_vel = 0
        elif self.y <= 0 and self.y_vel < 0:
            self.y_vel = 0

        speed = math.sqrt(self.x_vel ** 2 + self.y_vel ** 2)
        if speed < 1.0:
            speed = 1.0
        speed = self.speed / speed
        next_x = self.x + self.x_vel * speed
        next_y = self.y + self.y_vel * speed
        
        old_x, old_y = self.x, self.y
        self.x = next_x
        self.y = next_y
        if self._check_collisions(game_state):
            self.x, self.y = old_x, old_y

        self.cooldown = max(0, self.cooldown - 1)

    def get_coordinates(self):
        return int(self.x), int(self.y)
    
    def fire_gun(self, window, game_config, game_state):
        """Fires whichever weapon that blockshead is using at the moment"""
        self.bonus_score = 0
        if self.cooldown == 0:
            shot = None
            if self.weapon == "pistol":
                shot = Pistol(game_config, game_state)
                self.cooldown = 30
            elif self.ammo() > 0:
                if self.weapon == "uzi":
                    shot = Uzi(game_config, game_state)
                    self.cooldown = 12
                    self.ammo_dict["uzi"] = max(0, self.ammo() - 1)
                elif self.weapon == "shotgun":
                    shot = Shotgun(window, game_state)
                    self.ammo_dict["shotgun"] = max(0, self.ammo() - 1)
                    self.cooldown = 45
                elif self.weapon == "fireball":
                    #shot = Fireball(window, game_config, game_state)
                    self.cooldown = 45
                elif self.weapon == "fake walls":
                    xvel, yvel = self.direction.value
                    coords = self.x - xvel * self.width + 1, self.y - yvel * (self.height+10) + 1
                    wall = Fakewall(game_config, init_coords=coords)
                    collisions = False
                    for zombie in game_state.zombies + game_state.devils + game_state.fakewalls:
                        if check_collision(wall, zombie):
                            collisions = True
                    if not collisions:
                        logging.info("fakewall added")
                        game_state.fakewalls.append(wall)
                        self.ammo_dict["fake walls"] = max(0, self.ammo() - 1)
                        self.cooldown = 15
                    else:    
                        logging.debug("fakewall not possible here")
                
                # Automatically change to pistol if out of ammo
                if self.ammo() == 0:
                    self.weapon = "pistol"
            else:
                logging.info("Out of ammo")

            return shot

    def injure(self, damage, game_state):
        self.health -= damage

class Zombie(object):
    def __init__(self, window, game_config, init_coords=None):
        self.direction = Direction.UP
        self.images = {}
        for direction in Direction:
            # 8 different devil images
            self.images[direction] = pygame.image.load(f"images/zombies/z{direction.name.lower()}.png")

        # pick a random starting point on the right side of the field. Zombies start on the left half.
        self.radius = 35
        self.width = 2 * self.radius
        self.height = 2 * self.radius

        if init_coords is None:
            self.x = random.randrange(self.radius, game_config.width // 20)
            self.y = random.randrange(self.radius, game_config.height - self.radius)
        else:
            self.x, self.y = init_coords
        
        # Rolling average location
        self.x_average, self.y_average = self.x, self.y

        self.speed = game_config.zombie_speed
        self.health = 50
        self.cooldown = 0
        self.injury_cooldown = 0
        self.multiplier = 0.5
        self.angles = [0, 45, -45, 90, -90]

        # Randomize movement on enemy level
        if random.choice([True, False, False]):
            self.angles = [20] + self.angles
        elif random.choice([True, False]):
            self.angles = [-20] + self.angles
    
    def _check_collisions(self, game_state, game_config=None):
        if game_config is not None:
            if self.x >= game_config.width or self.x < 0:
                return True
            if self.y >= game_config.height or self.y < 0:
                return True

        for zombie in game_state.zombies + game_state.devils + game_state.fakewalls + [game_state.blockshead]:
            if zombie is self:
                continue
            if check_collision(self, zombie):
                return True
        return False

    def move(self, window, game_config, game_state):
        target = game_state.blockshead
        
        self.x_average = self.x_average * 0.9 + self.x * 0.1
        self.y_average = self.y_average * 0.9 + self.y * 0.1
        self.stuck = math.sqrt((self.x_average -self.x) ** 2 + (self.y_average - self.y) ** 2) <= 3
        if self.cooldown > 0:
            self.cooldown -= 1
            return
        if self.injury_cooldown > 0:
            self.injury_cooldown -= 1
            return

        diff = np.array([self.x - target.x, self.y - target.y])
        diff = diff / np.linalg.norm(diff) * self.speed
        
        dotprod, best_diff = -100.0, Direction.UP.value
        for d in Direction:
            d_val = d.value / np.linalg.norm(d.value)
            matching = np.dot(d_val, diff)
            if matching >= dotprod:
                dotprod = matching
                best_diff = d_val
                
        diff = best_diff * self.speed
        
        old_x, old_y = self.x, self.y
        for angle in self.angles:
            theta = np.radians(angle)
            c, s = np.cos(theta), np.sin(theta)
            R = np.array(((c, -s), (s, c)))
            direction = R @ diff
            next_x = self.x - direction[0]
            next_y = self.y - direction[1]

            self.x = next_x
            self.y = next_y
            collisions = self._check_collisions(game_state, game_config)
            if collisions:
                self.x, self.y = old_x, old_y
            else:
                # Calculate which direction is the best aligned with where the zombie ends up going
                dirs = [(d_prime, d_prime.value / np.linalg.norm(d_prime.value)) for d_prime in Direction]
                dirs_dots = [(d_prime, np.dot(-direction, d_prime_val)) for d_prime, d_prime_val in dirs]
                self.direction = max(dirs_dots, key=lambda t: t[1])[0]

                # Exit loop if no collisions are found
                break
                    
        # Cooldown period for shooting
        if self.cooldown > 0:
            self.cooldown -= 1
        if self.injury_cooldown > 0:
            self.injury_cooldown -= 1

    def get_coordinates(self):
        return int(self.x), int(self.y)

    def get_image(self):            
        return self.images[self.direction]

    def contact(self, game_state):
        bh = game_state.blockshead
        # Only attack walls if they are directly in the way of blockshead
        bh_diff = np.array([bh.x - self.x, bh.y - self.y])
        bh_diff = bh_diff / np.linalg.norm(bh_diff)
        
        targets = [game_state.blockshead]
        if self.stuck:
            targets += game_state.fakewalls
        for target in targets:
            target_diff = np.array([target.x - self.x, target.y - self.y])
            target_diff = target_diff / np.linalg.norm(target_diff)
            cosine = np.dot(target_diff, bh_diff)
            
            # Note that the cosine is always 1.0 if target == blockshead
            if not cosine >= 0.95:
                continue

            horizontal = abs(target.x - self.x) < target.width // 2 + self.radius + 2
            vertical = abs(target.y - self.y) < target.height // 2 + self.radius + 2
            
            if horizontal and vertical and self.cooldown == 0:
                target.injure(1, game_state)
                self.cooldown = 20
                break
    
    def injure(self, damage, game_state, direction=None):
        self.health -= damage
        self.injury_cooldown = 11
        if direction is not None:
            old_x, old_y = self.x, self.y
            self.x += direction.value[0] * 5
            self.y += direction.value[1] * 5
            collisions = self._check_collisions(game_state)
            if collisions:
                self.x, self.y = old_x, old_y

        if self.health <= 1:
            blood_mark = Blood(self.x, self.y)
            game_state.blood_marks.append(blood_mark)
            game_state.score += int(5 * game_state.multiplier)
            game_state.multiplier += self.multiplier

class Devil(object):
    def __init__(self, window, game_config):
        self.direction = Direction.UP
        self.images = {}
        for direction in Direction:
            # 8 different devil images
            self.images[direction] = pygame.image.load(f"images/devils/d{direction.name.lower()}.png")

        # pick a random starting point on the right side of the field. Zombies start on the left half.
        self.radius = 35
        self.width = 2 * self.radius
        self.height = 2 * self.radius
        
        self.x = random.randrange(self.radius, game_config.width // 4)
        self.y = random.randrange(self.radius, game_config.height - self.radius)
        self.x_average = self.x
        self.y_average = self.y
        
        self.speed = game_config.devil_speed
        self.health = 150
        self.cooldown = 0
        self.injury_cooldown = 0
        self.multiplier = 0.5
        self.angles = [0, 30, -30, 60, -60, 90, -90, 110, -110, 130, -130]

        # Randomize movement on enemy level
        if random.choice([True, False, False]):
            self.angles = [20] + self.angles
        elif random.choice([True, False]):
            self.angles = [-20] + self.angles
    
    def _check_collisions(self, game_state, game_config=None):
        if game_config is not None:
            if self.x >= game_config.width or self.x < 0:
                return True
            if self.y >= game_config.height or self.y < 0:
                return True

        for zombie in game_state.zombies + game_state.devils + game_state.fakewalls + [game_state.blockshead]:
            if zombie is self:
                continue
            if check_collision(self, zombie):
                return True
        return False

    def move(self, window, game_config, game_state):
        target = game_state.blockshead
        if self.cooldown > 0:
            self.cooldown -= 1
        if self.injury_cooldown > 0:
            self.injury_cooldown -= 1
            return

        diff = np.array([self.x - target.x, self.y - target.y])
        diff = diff / np.linalg.norm(diff) * self.speed
        
        old_x, old_y = self.x, self.y
        for angle in self.angles:
            theta = np.radians(angle)
            c, s = np.cos(theta), np.sin(theta)
            R = np.array(((c, -s), (s, c)))
            direction = R @ diff
            self.x = self.x - direction[0]
            self.y = self.y - direction[1]

            collisions = self._check_collisions(game_state, game_config)
            if collisions:
                self.x, self.y = old_x, old_y
            else:
                break
                    
        # Cooldown period for shooting
        if self.cooldown > 0:
            self.cooldown -= 1
        if self.injury_cooldown > 0:
            self.injury_cooldown -= 1
            
        self.x_average = self.x_average * 0.9 + self.x * 0.1
        self.y_average = self.y_average * 0.9 + self.y * 0.1

    def get_coordinates(self):
        return int(self.x), int(self.y)

    def get_image(self):            
        return self.images[self.direction]

    def contact(self, game_state):
        target = game_state.blockshead
        horizontal = abs(target.x - self.x) < target.width // 2 + self.radius + 2
        vertical = abs(target.y - self.y) < target.height // 2 + self.radius + 2
        if horizontal and vertical and self.cooldown == 0 and self.injury_cooldown == 0:
            target.injure(1, game_state)
            self.cooldown = 20
        elif self.cooldown == 0 and self.injury_cooldown == 0:
            logging.debug("Devil attack")
            attack = DevilAttack(self, game_state)
            game_state.shots.append(attack)
            self.cooldown = 300
            self.injury_cooldown = 70
    
    def injure(self, damage, game_state, direction=None):
        self.health -= damage
        self.injury_cooldown = 11
        if direction is not None:
            old_x, old_y = self.x, self.y
            self.x += direction.value[0] * 5
            self.y += direction.value[1] * 5
            collisions = self._check_collisions(game_state)
            if collisions:
                self.x, self.y = old_x, old_y

        if self.health <= 1:
            blood_mark = Blood(self.x, self.y)
            healthbox = Healthbox(game_state, self.x, self.y)
            game_state.blood_marks.append(blood_mark)
            game_state.healthboxes.append(healthbox)
            game_state.score += int(5 * game_state.multiplier)
            game_state.multiplier += self.multiplier
